ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/classic/JmsXAResourceRecoveryHelper.java

```diff
@@ -0,0 +1,178 @@
+            Logger.getLogger(JmsXAResourceRecoveryHelper.class);
+
+    private final XAConnectionFactory xaConnectionFactory;
+
+    private XAConnection xaConnection;
+
+    public JmsXAResourceRecoveryHelper(XAConnectionFactory xaConnectionFactory) {
+        this.xaConnectionFactory = xaConnectionFactory;
+    }
+
+    @Override
+    public boolean initialise(String properties) {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Initialise with properties=" + properties);
+        }
+
+        return true;
+    }
+
+    @Override
+    public XAResource[] getXAResources() {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Returning XA resource: " + this);
+        }
+
+        return new XAResource[] { this };
+    }
+
+    @Override
+    public void start(Xid xid, int i) throws XAException {
+        try {
+            getDelegate().start(xid, i);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public void end(Xid xid, int i) throws XAException {
+        try {
+            getDelegate().end(xid, i);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public int prepare(Xid xid) throws XAException {
+        try {
+            return getDelegate().prepare(xid);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public void commit(Xid xid, boolean b) throws XAException {
+        try {
+            getDelegate().commit(xid, b);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public void rollback(Xid xid) throws XAException {
+        try {
+            getDelegate().rollback(xid);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public Xid[] recover(int i) throws XAException {
+        try {
+            return getDelegate().recover(i);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public boolean isSameRM(XAResource xaResource) throws XAException {
+        try {
+            return getDelegate().isSameRM(xaResource);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public void forget(Xid xid) throws XAException {
+        try {
+            getDelegate().forget(xid);
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public int getTransactionTimeout() throws XAException {
+        try {
+            return getDelegate().getTransactionTimeout();
+        } finally {
+            disconnect();
+        }
+    }
+
+    @Override
+    public boolean setTransactionTimeout(int i) throws XAException {
+        try {
+            return getDelegate().setTransactionTimeout(i);
+        } finally {
+            disconnect();
+        }
+    }
+
+    private XAResource getDelegate() throws XAException {
+        try {
+            xaConnection = xaConnectionFactory.createXAConnection();
```

Comment:
as discussed on issue, lets cache this and release it during recover(TMENDRSCAN)